runtime/tr.source/trj9/z/codegen/J9S390CHelperLinkage.cpp

```diff
@@ -252,13 +260,14 @@ class RealRegisterManager
  *    \param callNode The node for which you are generating a heleper call
  *    \param deps The pre register dependency conditions that will be filled by this function to attach within ICF
  *    \param paramInRegisters Stack of registers that contains arguments to be passed to the helper function,
- *           if passed empty stack, this function will evaluate children of node to prepare arguments
+ *           if passed empty stack, this function will clobber evaluate children of node to prepare arguments
  *    \return TR::Register *helperReturnResult, gets the return value of helper function and return to the evaluator. 
  */
-TR::Register * TR::S390CHelperLinkage::buildDirectDispatch(TR::Node * callNode, TR::RegisterDependencyConditions **deps, TR_Stack<TR::Register*>& paramInRegisters)
+TR::Register * TR::S390CHelperLinkage::buildDirectDispatch(TR::Node * callNode, TR::RegisterDependencyConditions **deps, TR_Stack<TR::Register*>& paramInRegisters, TR::Register *returnReg)
```

Comment:
Is `paramInRegisters` even being used anywhere in the codebase? I don't see any occurrences. Is it possible to simplify this API by removing the parameter?

Reply:
Currently paramInRegister is used only by VMNewEvaluator. In cases where we have child iconst or iload (Basically 32-bit values to be passed as argument), evaluating that child only changes the lower half of the register keeping upper half intact. This was hitting an issue where C helper function was trying to use whole argument register.

Reply:
Is it possible to remove the use and deprecate this parameter?

Reply:
In VMNew evaluator with current implementation of `iconstEval*` on it is not possible it uses `LHI` instruction to load so we end up having only lower 32 bits of the register filled with value. `new` helper uses whole 64 bits in argument. This causes allocation of wrong type of object or very large size array which causes `AIOOB`.

Reply:
Ok, I understand. Still seems odd that we have a very general API for a very specific case. IMO we would benefit from having consistency everywhere and sending hand crafted call nodes to the linkage, same as Power codegen does. In this would you can have an `i2l` tree for the scenario you describe in VMNewEvaluator and everything just works and is consistent.

This is outside the scope of this PR though, but a nice to have in the future.